orangecontrib/geo/widgets/owmap.py

```diff
@@ -1,906 +1,849 @@
         """
-        @staticmethod
-        def latlon_to_easting_northing(lat, lon):
-            R = 6378137
-            MAX_LATITUDE = 85.0511287798
-            DEG = np.pi / 180
-
-            lat = np.clip(lat, -MAX_LATITUDE, MAX_LATITUDE)
-            sin = np.sin(DEG * lat)
-            x = R * DEG * lon
-            y = R / 2 * np.log((1 + sin) / (1 - sin))
-            return x, y
-
-        @staticmethod
-        def easting_northing_to_pixel(x, y, zoom_level, pixel_origin, map_pane_pos):
-            R = 6378137
-            PROJ_SCALE = .5 / (np.pi * R)
-
-            zoom_scale = 256 * (2 ** zoom_level)
-            x = (zoom_scale * (PROJ_SCALE * x + .5)).round() + (map_pane_pos[0] - pixel_origin[0])
-            y = (zoom_scale * (-PROJ_SCALE * y + .5)).round() + (map_pane_pos[1] - pixel_origin[1])
-            return x, y
-
-    N_POINTS_PER_ITER = 1500
-
-    def redraw_markers_overlay_image(self, *args, new_image=False):
-        if not args and not self._drawing_args or self.data is None:
+        if t.n_loadings == 3:
+            self._show_internet_error(True)
             return
 
-        if args:
-            self._drawing_args = args
-        north, east, south, west, width, height, zoom, origin, map_pane_pos = self._drawing_args
-
-        lat, lon = self._latlon_data.T
-        visible = ((lat <= north) & (lat >= south) &
-                   (lon <= east) & (lon >= west)).nonzero()[0]
-        in_subset = (np.in1d(self.data.ids, self._subset_ids)
-                     if self._subset_ids.size else
-                     np.tile(True, len(lon)))
-
-        is_js_path = self.is_js_path = len(visible) < self.N_POINTS_PER_ITER
-
-        self._update_legend(is_js_path)
+        future = self.loader.get(t)
+        @future.add_done_callback
+        def set_tile(_future):
+            if _future.cancelled():
+                return
 
-        np.random.shuffle(visible)
-        # Sort points in subset to be painted last
-        visible = visible[np.lexsort((in_subset[visible],))]
+            assert _future.done()
 
-        if is_js_path:
-            self.evalJS('clear_markers_overlay_image()')
-            self._update_js_markers(visible, in_subset[visible])
-            self._owwidget.disable_some_controls(False)
-            return
+            if _future.exception():
+                _t = _future.exception().tile
+                _t.n_loadings += 1
+                # retry to download image
+                self._load_one_from_net(_t)
+            else:
+                _t, img = _future.result()
+                if not _t.disc_cache:
+                    self._show_internet_error(False)
+                self.mem_cache[_t] = img
+                self._add_patch(_t)
+                self._update_map_item()
+
+        self.futures.append(future)
+
+    def _show_internet_error(self, show):
+        if not self.master.Warning.no_internet.is_shown() and show:
```

Comment:
I'd rather you didn't refer to master's warnings like that. What if (hypothetical) master doesn't have this warning? It would be better to emit a signal here, and handle it on the master.